internal/crypto/crypto.go

```diff
@@ -299,27 +297,24 @@ func (k *Key) Decrypt(plaintext []byte, ciphertextWithMac []byte) (int, error) {
 	l := len(ciphertextWithMac) - macSize
 	ciphertextWithIV, mac := ciphertextWithMac[:l], ciphertextWithMac[l:]
 
+	// extract iv
+	iv, ciphertext := ciphertextWithIV[:ivSize], ciphertextWithIV[ivSize:]
+
 	// verify mac
-	if !poly1305Verify(ciphertextWithIV[ivSize:], ciphertextWithIV[:ivSize], &k.MACKey, mac) {
+	if !poly1305Verify(ciphertext, iv, &k.MACKey, mac) {
 		return 0, ErrUnauthenticated
 	}
 
-	// extract iv
-	iv, ciphertext := ciphertextWithIV[:ivSize], ciphertextWithIV[ivSize:]
-
 	if len(ciphertext) != plaintextLength {
-		return 0, errors.Errorf("plaintext and ciphertext lengths do not match: %d != %d", len(ciphertext), plaintextLength)
+		panic("plaintext and ciphertext lengths do not match")
 	}
 
 	// decrypt data
 	c, err := aes.NewCipher(k.EncryptionKey[:])
 	if err != nil {
 		panic(fmt.Sprintf("unable to create cipher: %v", err))
 	}
-
-	// decrypt
 	e := cipher.NewCTR(c, iv)
-	plaintext = plaintext[:len(ciphertext)]
```

Comment:
What happens if `len(plaintext)` is greater than `len(ciphertext)-Extension`? I think this line needs to stay, what do you think?

Reply:
`XORKeyStream` is defined in that case to only operate on `len(dst)`.

I am, to be honest, not a fan of this API because it does not return the plaintext slice, only the length, and I can see an user failing to slice the plaintext in such a case, leaving trailing unauthenticated data.